src/test/jenkins/jenkinsfiles/Declarative_thisObject_JenkinsFile

```diff
@@ -0,0 +1,51 @@
+pipeline {
+
+    agent any
+
+    environment {
+        CC = testThis(this, 'gcc')
+    }
+
+    options {
+        buildDiscarder(logRotator(numToKeepStr: testThis(this, '10')))
+    }
+
+    triggers {
+        pollSCM(testThis(this, '*/5 * * * *'))
+    }
+
+    stages {
+        stage('Checkout') {
+            when {
+                environment name: 'CC', value: testThis(this, 'clang')
+            }
+            steps {
+                echo testThis(this, "Steps")
+            }
+        }
+
+        stage('build') {
+            agent { docker testThis(this, 'maven:3-alpine') }
+            options {
+                timeout(time: 20, unit: 'MINUTES')
+            }
+            steps {
+                withEnv(["GRADLE_HOME=${tool name: 'GRADLE_3', type: 'hudson.plugins.gradle.GradleInstallation'}"]) {
+                    echo testThis(this, "In env")
+                }
+            }
+        }
+    }
+
+    post {
+        always {
+            echo testThis(this, 'pipeline unit tests completed')
+        }
+
+    }
+}
+
+String testThis(Script scriptRef, String text){
+    echo "This is a script?: ${this == scriptRef}"
+    return text
```

Comment:
return "$text: This is a script?: ${this == scriptRef}"

Would allow to:
assertCallStack().contains("pipeline unit tests completed: This is a script?: true")

So you can help other contributors to detect broken places in the code in later updates

Reply:
Then we could only call the method in 1 place. I called it from environment, options, when, trigger and script content now.

Reply:
Scratch that, it technically works, but the values returned would not be valid in real life. For test it is okay i think

Reply:
I updated it a little to pass in Object. So that this always works and we can check instanceof

Reply:
In addition to that i added an assert so that it may not contain false